focus/internals/src/lib/project_cache/mod.rs

```diff
@@ -0,0 +1,430 @@
+            .optional_projects
+            .underlying
+            .iter()
+            .map(|(name, _project)| name.clone())
+            .collect::<BTreeSet<String>>();
+        let project_names = Vec::from_iter(project_names.into_iter());
+        let (build_graph_hash_key, build_graph_hash) =
+            self.get_build_graph_hash(commit_id, true)?;
+        let mut keys = vec![build_graph_hash_key];
+        let shard_size = project_names.len() / shard_count;
+        let mut shards = project_names.chunks(shard_size).skip(shard_index);
+        let build_graph_hash_str = hex::encode(&build_graph_hash);
+        let commit_id_str = hex::encode(commit_id.as_bytes());
+        let project_names = shards
+            .next()
+            .ok_or_else(|| anyhow::anyhow!("No such shard"))?;
+        info!(
+            ?shard_index,
+            ?shard_count,
+            ?project_names,
+            commit_id = ?commit_id_str,
+            build_graph_hash = ?build_graph_hash_str,
+            "Generating project pattern cache"
+        );
+        for project_name in project_names {
+            let (key, value) = self
+                .get_project(commit_id, &build_graph_hash, project_name, true)
+                .with_context(|| {
+                    format!("Generating cache content for project {}", project_name)
+                })?;
+            let _value = value.ok_or_else(|| {
+                anyhow::anyhow!(
+                    "Could not generate patterns to cache for project '{}'",
+                    project_name
+                )
+            })?;
+            keys.push(key);
+        }
+
+        Ok(keys)
+    }
+
+    /// Get (possibly fault) a project from the cache.
+    pub fn get_project(
+        &self,
+        commit_id: Oid,
+        build_graph_hash: &[u8],
+        project_name: &str,
+        fault: bool,
+    ) -> anyhow::Result<(NamespacedKey, Option<Value>)> {
+        let calculate = move |key: &Key, repo: &Repo| -> anyhow::Result<Option<Value>> {
+            if let Key::OptionalProjectPatternSet { project_name, .. } = key {
+                let selection_manager = repo.selection_manager()?;
+                let catalog = selection_manager.project_catalog();
+                let span = info_span!("Resolving");
+                let _guard = span.enter();
+
+                let project = catalog
+                    .optional_projects
+                    .underlying
+                    .get(project_name)
+                    .ok_or_else(|| {
+                        anyhow::anyhow!(format!("No such project '{}'", &project_name))
+                    })?;
+                let targets = {
+                    let mut targets = HashSet::<Target>::new();
+                    for repr in &project.targets {
+                        let target = Target::try_from(repr.as_str())?;
+                        targets.insert(target);
+                    }
+                    targets
+                };
```

Comment:
This doesn't allow projects that include sub-project targets, since I wasn't able to continue the work to get the project target resolution into try_from.

If you want all the recursive targets from a project that potentially includes other projects:
```
let targets = project.resolve_targets_for_project(catalog.optional_projects.underlying)
```

Reply:
> Another thought: What's the purpose of having the intermediate buildgraph hash key. Why not go from Commit to Projects? (Maybe this is answered in the code somewhere and I'm trying to answer it myself now)

Added a tracking issue: https://github.com/twitter/focus/issues/24